splay/src/main/scala/org/scalajs/benchmark/splay/Splay.scala

```diff
@@ -0,0 +1,439 @@
+      splay(key)
+      if (root.key != key) {
+        val node = new Node(key, value)
+        if (key > root.key) {
+          node.left = root
+          node.right = root.right
+          root.right = null
+        } else {
+          node.right = root
+          node.left = root.left
+          root.left = null
+        }
+        root = node
+      }
+    }
+  }
+
+  /**
+   * Removes a node with the specified key from the tree if the tree
+   * contains a node with this key. The removed node is returned. If the
+   * key is not found, an exception is thrown.
+   *
+   * @param {number} key Key to find and remove from the tree.
+   * @return {SplayTree.Node} The removed node.
+   */
+  def remove(key: Int): Node = {
+    if (isEmpty) {
+      throw new Exception(s"Key not found: $key")
+    } else {
+      splay(key)
+      if (root.key != key) {
+        throw new Exception(s"Key not found: $key")
+      } else {
+        val removed = root
+        if (root.left == null) {
+          root = root.right
+        } else {
+          val right = root.right
+          root = root.left
+          // Splay to make sure that the new root has an empty right child.
+          splay(key)
+          // Insert the original right child as the right child of the new
+          // root.
+          root.right = right
+        }
+        removed
+      }
+    }
+  }
+
+  /**
+   * Returns the node having the specified key or null if the tree doesn't contain
+   * a node with the specified key.
+   *
+   * @param {number} key Key to find in the tree.
+   * @return {SplayTree.Node} Node having the specified key.
+   */
+  def find(key: Int): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      splay(key)
+      if (root.key == key) root
+      else null
+    }
+  }
+
+  /**
+   * @return {SplayTree.Node} Node having the maximum key value.
+   */
+  def findMax(opt_startNode: Node): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      var current =
+        if (opt_startNode != null) opt_startNode
+        else root
+      while (current.right != null) {
+        current = current.right
+      }
+      current
+    }
+  }
+
+  /**
+   * @return {SplayTree.Node} Node having the maximum key value that
+   *     is less than the specified key value.
+   */
+  def findGreatestLessThan(key: Int): Node = {
+    if (isEmpty) {
+      null
+    } else {
+      // Splay on the key to move the node with the given key or the last
+      // node on the search path to the top of the tree.
+      splay(key)
+      // Now the result is either the root node or the greatest node in
+      // the left subtree.
+      if (root.key < key) {
+        root
+      } else if (root.left != null) {
+        findMax(root.left)
+      } else {
+        null
+      }
+    }
+  }
+
+  /**
+   * @return {Array<*>} An array containing all the keys of tree's nodes.
+   */
+  def exportKeys: Array[Int] = {
+    // Both variants have a similar performance
```

Comment:
Yes, because `Array.newBuilder` is *aggressively* optimized away as basically the same as the other guy ^^